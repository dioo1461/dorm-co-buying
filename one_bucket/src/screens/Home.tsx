import strings from '@/constants/strings'
import { AppContext } from '@/hooks/useContext/AppContext'
import { useNavigation } from '@react-navigation/native'
import React, { useContext, useRef, useState } from 'react'
import {
    Dimensions,
    FlatList,
    ScrollView,
    StyleSheet,
    Text,
    TouchableOpacity,
    View,
} from 'react-native'

const { width: SCREEN_WIDTH, height: SCREEN_HEIGHT } = Dimensions.get('window')

const Home: React.FC = (): React.JSX.Element => {
    
    
    const navigation = useNavigation()
    const { themeColor } = useContext(AppContext)
    const [data, setData] = useState([
        'post1',
        'post2',
        'post3',
        'post4',
        'post5',
        'post6',
        'post7',
        'post8',
        'post9',
    ])
    const flatListRef = useRef<FlatList>(null)

    const renderItem = ({ item }: { item: string }) => (
        <View style={styles.post}>
            <Text>{item}</Text>
        </View>
    )

    const onScrollEnd = (event: any) => {
        const offsetX = event.nativeEvent.contentOffset.x
        const contentWidth = event.nativeEvent.contentSize.width
        if (offsetX >= contentWidth - SCREEN_WIDTH) {
            flatListRef.current?.scrollToOffset({ animated: false, offset: 0 })
        }
    }

    const [board, setBoard] = useState(0);
    const home = () => setBoard(0);
    const trade = () => setBoard(1);
    const used = () => setBoard(2);
    const free = () => setBoard(3);

    const showBoard = () => {
        if (board === 0) return (
            <View>
                <View style={styles.title}>
                        <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                            추천 거래글
                        </Text>
                    </View>
                    <FlatList
                        ref={flatListRef}
                        data={data}
                        renderItem={renderItem}
                        keyExtractor={(item, index) => index.toString()}
                        horizontal
                        pagingEnabled
                        showsHorizontalScrollIndicator={false}
                        onMomentumScrollEnd={onScrollEnd}
                        contentContainerStyle={styles.post_list}
                    />
                    <View style={styles.title}>
                        <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                            내가 참여한 거래글
                        </Text>
                    </View>
                    <FlatList
                        ref={flatListRef}
                        data={data}
                        renderItem={renderItem}
                        keyExtractor={(item, index) => index.toString()}
                        horizontal
                        pagingEnabled
                        showsHorizontalScrollIndicator={false}
                        onMomentumScrollEnd={onScrollEnd}
                        contentContainerStyle={styles.post_list}
                    />
                    <View style={styles.title}>
                        <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                            내가 쓴 거래글
                        </Text>
                    </View>
                    <FlatList
                        ref={flatListRef}
                        data={data}
                        renderItem={renderItem}
                        keyExtractor={(item, index) => index.toString()}
                        horizontal
                        pagingEnabled
                        showsHorizontalScrollIndicator={false}
                        onMomentumScrollEnd={onScrollEnd}
                        contentContainerStyle={styles.post_list}
                    />
            </View>
        )
        if (board === 1) return (
            <ScrollView pagingEnabled contentContainerStyle={styles.post_list_vert}>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post1</Text></View>
                <View style={styles.post_thumb}><Text>img1</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post2</Text></View>
                <View style={styles.post_thumb}><Text>img2</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post3</Text></View>
                <View style={styles.post_thumb}><Text>img3</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post4</Text></View>
                <View style={styles.post_thumb}><Text>img4</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post5</Text></View>
                <View style={styles.post_thumb}><Text>img5</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post6</Text></View>
                <View style={styles.post_thumb}><Text>img6</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post7</Text></View>
                <View style={styles.post_thumb}><Text>img7</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post8</Text></View>
                <View style={styles.post_thumb}><Text>img8</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post9</Text></View>
                <View style={styles.post_thumb}><Text>img9</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post10</Text></View>
                <View style={styles.post_thumb}><Text>img10</Text></View>
                </View>
            </ScrollView>
        )
        if (board === 2) return (
            <ScrollView pagingEnabled contentContainerStyle={styles.post_list_vert}>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post1</Text></View>
                <View style={styles.post_thumb}><Text>img1</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post2</Text></View>
                <View style={styles.post_thumb}><Text>img2</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post3</Text></View>
                <View style={styles.post_thumb}><Text>img3</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post4</Text></View>
                <View style={styles.post_thumb}><Text>img4</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post5</Text></View>
                <View style={styles.post_thumb}><Text>img5</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post6</Text></View>
                <View style={styles.post_thumb}><Text>img6</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post7</Text></View>
                <View style={styles.post_thumb}><Text>img7</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post8</Text></View>
                <View style={styles.post_thumb}><Text>img8</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post9</Text></View>
                <View style={styles.post_thumb}><Text>img9</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post10</Text></View>
                <View style={styles.post_thumb}><Text>img10</Text></View>
                </View>
            </ScrollView>
        )
        if (board === 3) return (
            <ScrollView pagingEnabled contentContainerStyle={styles.post_list_vert}>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post1</Text></View>
                <View style={styles.post_thumb}><Text>img1</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post2</Text></View>
                <View style={styles.post_thumb}><Text>img2</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post3</Text></View>
                <View style={styles.post_thumb}><Text>img3</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post4</Text></View>
                <View style={styles.post_thumb}><Text>img4</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post5</Text></View>
                <View style={styles.post_thumb}><Text>img5</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post6</Text></View>
                <View style={styles.post_thumb}><Text>img6</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post7</Text></View>
                <View style={styles.post_thumb}><Text>img7</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post8</Text></View>
                <View style={styles.post_thumb}><Text>img8</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post9</Text></View>
                <View style={styles.post_thumb}><Text>img9</Text></View>
                </View>
                <View style={styles.post_vert}>
                <View style={styles.post_cont}><Text>post10</Text></View>
                <View style={styles.post_thumb}><Text>img10</Text></View>
                </View>
            </ScrollView>
        )
        return null;
    }

    return (
        <>
            <View style={{ flex: 1 }}>
                <View style={{ flex: 1, backgroundColor: 'white' }}>
                    <View style={styles.menu}>
                        <TouchableOpacity onPress={home}>
                            <View style={{...styles.menu_S, backgroundColor: board==0 ? "yellow" : "white"}}>
                                <Text>홈</Text>
                            </View>
                        </TouchableOpacity>
                        <TouchableOpacity onPress={trade}>
                            <View style={{...styles.menu_S, backgroundColor: board==1 ? "yellow" : "white"}}>
                                <Text>거래게시판</Text>
                            </View>
                        </TouchableOpacity>
                        <TouchableOpacity onPress={used}>
                            <View style={{...styles.menu_S, backgroundColor: board==2 ? "yellow" : "white"}}>
                                <Text>중고게시판</Text>
                            </View>
                        </TouchableOpacity>
                        <TouchableOpacity onPress={free}>
                            <View style={{...styles.menu_S, backgroundColor: board==3 ? "yellow" : "white"}}>
                                <Text>자유게시판</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                    <View>{showBoard()}</View>
                </View>
            </View>
            <TouchableOpacity
                style={[styles.fab, { backgroundColor: themeColor.ICON_BG }]}
                onPress={() =>
                    navigation.navigate(strings.postGroupPurchaseScreenName)
                }>
                <Text style={styles.fabIcon}>+</Text>
            </TouchableOpacity>
        </>
    )
}

const styles = StyleSheet.create({
    menu: {
        height: 50,
        backgroundColor: 'gray',
        flexDirection: 'row',
    },
    menu_S: {
        height: 50,
        width: SCREEN_WIDTH / 4,
        backgroundColor: 'white',
        justifyContent: 'center',
        alignItems: 'center',
    },
    title: {
        height: 30,
        backgroundColor: 'white',
        marginStart: 14,
    },
    post_list: {
        backgroundColor: 'white',
    },
    post: {
        width: SCREEN_WIDTH / 3,
        height: (SCREEN_WIDTH / 3) * 1.3,
        backgroundColor: 'white',
        justifyContent: 'center',
        alignItems: 'center',
    },
    post_list_vert:{
        backgroundColor: "white",
      },
      post_vert:{
        height: SCREEN_HEIGHT/8,
        backgroundColor: "blue",
        flexDirection: "row",
      },
      post_cont:{
        height: SCREEN_HEIGHT/8,
        width: (SCREEN_WIDTH)-(SCREEN_HEIGHT/8),
        backgroundColor: "white",
        justifyContent: "center",
      },
      post_thumb:{
        height: SCREEN_HEIGHT/8,
        width: SCREEN_HEIGHT/8,
        backgroundColor: "gray",
        justifyContent: "center",
        alignItems: "center",
      },
    fab: {
        position: 'absolute',
        width: 56,
        height: 56,
        alignItems: 'center',
        justifyContent: 'center',
        right: 30,
        bottom: 30,
        borderRadius: 30,
        elevation: 8,
    },
    fabIcon: {
        fontSize: 40,
        color: 'white',
    },
})

export default Home
