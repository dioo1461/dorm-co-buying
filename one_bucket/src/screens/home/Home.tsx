import { darkColors, Icolor, lightColors } from '@/constants/colors'
import { useBoundStore } from '@/hooks/useStore/useBoundStore'
import { useEffect, useRef, useState } from 'react'
import {
    Appearance,
    Dimensions,
    FlatList,
    StyleSheet,
    Text,
    View,
} from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { stackNavigation } from '../navigation/NativeStackNavigation'

const { width: SCREEN_WIDTH, height: SCREEN_HEIGHT } = Dimensions.get('window')

const Home: React.FC = (): JSX.Element => {
    const { themeColor, setThemeColor } = useBoundStore(state => ({
        themeColor: state.themeColor,
        setThemeColor: state.setThemeColor,
    }))

    // 다크모드 변경 감지
    useEffect(() => {
        const themeSubscription = Appearance.addChangeListener(
            ({ colorScheme }) => {
                setThemeColor(colorScheme === 'dark' ? darkColors : lightColors)
            },
        )
        return () => themeSubscription.remove()
    }, [])

    const styles = createStyles(themeColor)
    const navigation = stackNavigation()
    const flatListRef = useRef<FlatList>(null)
    const [data, setData] = useState([
        'post1',
        'post2',
        'post3',
        'post4',
        'post5',
        'post6',
        'post7',
        'post8',
        'post9',
        'post10',
    ])

    const renderItem_Box = ({ item }: { item: string }) => (
        <TouchableOpacity style={styles.postBox}>
            <View style={styles.postBox_thumbFrame}>
                <View style={styles.postBox_thumb}>
                    <Text>{item}</Text>
                </View>
                <View style={styles.postBox_part}>
                    <Text style={{ color: 'black' }}>0/10</Text>
                </View>
            </View>
            <View style={styles.postBox_cont}>
                <Text>{item}</Text>
            </View>
        </TouchableOpacity>
    )

    const renderItem_Horz = ({ item }: { item: string }) => (
        <View style={styles.postHorz}>
            <View style={styles.postHorz_cont}>
                <Text>{item}</Text>
            </View>
            <View style={styles.postHorz_thumbFrame}>
                <View style={styles.postHorz_thumb}>
                    <Text>{item}</Text>
                </View>
            </View>
        </View>
    )

    const onScrollEnd = (event: any) => {
        const offsetX = event.nativeEvent.contentOffset.x
        const contentWidth = event.nativeEvent.contentSize.width
        if (offsetX >= contentWidth - SCREEN_WIDTH) {
            flatListRef.current?.scrollToOffset({ animated: false, offset: 0 })
        }
    }

    return (
        <View>
            <View style={styles.homeTitle}>
                <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                    추천 거래글
                </Text>
            </View>
            <FlatList
                ref={flatListRef}
                data={data}
                renderItem={renderItem_Box}
                keyExtractor={(item, index) => index.toString()}
                horizontal
                pagingEnabled
                showsHorizontalScrollIndicator={false}
                onMomentumScrollEnd={onScrollEnd}
                contentContainerStyle={styles.postList}
            />
            <View style={styles.homeTitle}>
                <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                    내가 참여한 거래글
                </Text>
            </View>
            <FlatList
                ref={flatListRef}
                data={data}
                renderItem={renderItem_Box}
                keyExtractor={(item, index) => index.toString()}
                horizontal
                pagingEnabled
                showsHorizontalScrollIndicator={false}
                onMomentumScrollEnd={onScrollEnd}
                contentContainerStyle={styles.postList}
            />
            <View style={styles.homeTitle}>
                <Text style={{ fontSize: 20, fontWeight: 'bold' }}>
                    내가 쓴 거래글
                </Text>
            </View>
            <FlatList
                key={'home'}
                ref={flatListRef}
                data={data}
                renderItem={renderItem_Box}
                keyExtractor={(item, index) => index.toString()}
                horizontal
                pagingEnabled
                showsHorizontalScrollIndicator={false}
                onMomentumScrollEnd={onScrollEnd}
                contentContainerStyle={styles.postList}
            />
            <TouchableOpacity
                style={styles.fab}
                onPress={() => navigation.navigate('CreateMarketPost')}>
                <Text style={styles.fabIcon}>+</Text>
            </TouchableOpacity>
        </View>
    )
}

const createStyles = (theme: Icolor) =>
    StyleSheet.create({
        upperFrame: {
            height: 50,
            flexDirection: 'row',
        },
        upperMenu: {
            height: 50,
            width: SCREEN_WIDTH / 4,
            justifyContent: 'center',
            alignItems: 'center',
            borderBottomWidth: 0.5,
        },
        homeTitle: {
            height: 30,
            marginStart: 14,
        },
        postList: {},
        postBox: {
            width: SCREEN_WIDTH / 3,
            height: (SCREEN_WIDTH / 3) * 1.3,
            alignItems: 'center',
        },
        postBox_thumbFrame: {
            width: SCREEN_WIDTH / 3,
            height: SCREEN_WIDTH / 3,
            justifyContent: 'center',
            alignItems: 'center',
        },
        postBox_thumb: {
            width: SCREEN_WIDTH / 3 - 5,
            height: SCREEN_WIDTH / 3 - 5,
            backgroundColor: 'lightgray',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 10,
            borderWidth: 0,
        },
        postBox_part: {
            width: 50,
            height: 20,
            marginTop: -20,
            marginLeft: SCREEN_WIDTH / 6 + 10,
            backgroundColor: 'lightblue',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 5,
            borderWidth: 0.3,
        },
        postBox_cont: {
            width: SCREEN_WIDTH / 3,
            height: (SCREEN_WIDTH / 3) * 0.3,
            paddingHorizontal: 10,
        },
        postHorz: {
            height: SCREEN_HEIGHT / 8,
            flexDirection: 'row',
        },
        postHorz_cont: {
            height: SCREEN_HEIGHT / 8,
            width: SCREEN_WIDTH - SCREEN_HEIGHT / 8,
            justifyContent: 'center',
            paddingHorizontal: 20,
        },
        postHorz_thumbFrame: {
            height: SCREEN_HEIGHT / 8,
            width: SCREEN_HEIGHT / 8,
            justifyContent: 'center',
            alignItems: 'center',
        },
        postHorz_thumb: {
            height: SCREEN_HEIGHT / 8 - 5,
            width: SCREEN_HEIGHT / 8 - 5,
            backgroundColor: 'lightgray',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 10,
            borderWidth: 0,
        },
        fab: {
            backgroundColor: theme.BUTTON_BG,
            position: 'absolute',
            width: 56,
            height: 56,
            alignItems: 'center',
            justifyContent: 'center',
            right: 25,
            bottom: 70,
            borderRadius: 30,
            elevation: 8,
        },
        fabIcon: {
            fontSize: 40,
            color: theme.BUTTON_TEXT,
        },
    })

export default Home
