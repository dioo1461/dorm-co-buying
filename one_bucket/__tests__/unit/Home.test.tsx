import { render } from '@testing-library/react-native'
import React from 'react'
import 'react-native'
import Home from 'screens/home/Home'

describe('Home', () => {
    it('renders screen Home', () => {
        const screen = render(<Home />)
        const json = screen.toJSON()
        expect(json).toMatchSnapshot()
    })

    it('displays title properly', () => {
        const screen = render(<Home />)
        const title = screen.getByText('home')
    })
})
