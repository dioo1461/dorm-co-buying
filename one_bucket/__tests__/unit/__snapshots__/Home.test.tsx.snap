// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders screen Home 1`] = `
<View>
  <Text>
    home
  </Text>
</View>
`;
